src/serve.rs

```diff
@@ -55,186 +51,221 @@ pub enum Error {
-    let args: Vec<ScVal> = if args_xdr.is_empty() {
-        args.iter()
-            .zip(inputs.iter())
-            .map(|(a, input)| strval::from_json(a, &input.type_))
-            .collect::<Result<Vec<_>, _>>()?
+    // TODO: Figure out and enforce the expected type here. For now, handle both a symbol and a
+    // binary. The cap says binary, but other implementations use symbol.
+    let method: String = if let ScVal::Object(Some(ScObject::Bytes(bytes))) = method_xdr {
+        bytes
+            .try_into()
+            .map_err(|_| Error::Xdr(XdrError::Invalid))?
+    } else if let ScVal::Symbol(bytes) = method_xdr {
+        bytes
+            .try_into()
+            .map_err(|_| Error::Xdr(XdrError::Invalid))?
     } else {
-        args_xdr
-            .iter()
-            .map(|a| match base64::decode(a) {
-                Err(_) => Err(StrValError::InvalidValue),
-                Ok(b) => ScVal::from_xdr(b).map_err(StrValError::Xdr),
-            })
-            .collect::<Result<Vec<_>, _>>()?
+        return Err(Error::Xdr(XdrError::Invalid));
     };
 
+    // Initialize storage and host
+    let ledger_entries = snapshot::read(ledger_file)?;
+
+    let snap = Rc::new(snapshot::Snap { ledger_entries });
+    let storage = Storage::with_recording_footprint(snap);
+    let h = Host::with_storage(storage);
+
+    // TODO: Check the parameters match the contract spec, or return a helpful error message
     let mut complete_args = vec![
         ScVal::Object(Some(ScObject::Bytes(contract_id.try_into()?))),
-        ScVal::Symbol(func.try_into()?),
+        ScVal::Symbol(method.try_into()?),
     ];
-    complete_args.extend_from_slice(args.as_slice());
+    complete_args.extend_from_slice(params);
 
     let res = h.invoke_function(HostFunction::Call, complete_args.try_into()?)?;
 
-    // TODO: Include costs in result struct
-    // let cost = h.get_budget(|b| {
-    //     let mut v = vec![
-    //         ("cpu_insns", b.cpu_insns.get_count()),
-    //         ("mem_bytes", b.mem_bytes.get_count()),
-    //     ];
-    //     // for cost_type in CostType::variants() {
-    //     //     v.push((cost_type.try_into()?, b.get_input(*cost_type)));
-    //     // }
-    //     Some(v)
-    // });
+    // Calculate the budget usage
+    let cost = h.get_budget(|b| {
+        let mut m = serde_json::Map::new();
+        m.insert(
+            "cpu_insns".to_string(),
+            Value::String(b.cpu_insns.get_count().to_string()),
+        );
+        m.insert(
+            "mem_bytes".to_string(),
+            Value::String(b.mem_bytes.get_count().to_string()),
+        );
+        // TODO: Include these extra costs. Figure out the rust type conversions.
+        // for cost_type in CostType::variants() {
+        //     m.insert(cost_type, b.get_input(*cost_type));
+        // }
+        m
+    });
 
+    // Calculate the storage footprint
     let (storage, _, _) = h.try_finish().map_err(|_h| {
```

Comment:
Instead of calling get_budget above, you should use the second parameter returned in `try_finish` which will be a `Budget`. See this code as an example:


https://github.com/stellar/soroban-cli/blob/13a51da46ee7d9553554f703ef28d83443367646/src/invoke.rs#L125-L137

Reply:
Oh cool. I think I copied it from an old version of invoke. Will update.